### DISCUSSION TITLE
Returning errors from templates

### DISCUSSION THREAD

**Discussion by tarndt** [a-h/templ#662] [Q&A] [answered]
The interface for componet is:

```go
type Component interface {
    // Render the template.
    Render(ctx context.Context, w io.Writer) error
}
```

but when writing a template the syntax is:
```
templ Name(....) {
  //markup here
}
```

What if I want ``Name`` to return a non-nil error? I have a template using an iterator that may fail:
```
row, ok, error := t.NextRow()
```

**Comment by a-h** [comment, marked as answer]
The idea was that you'd have created a view model already, and wouldn't be doing logic that can fail inside a template, just like http.Handler doesn't return an error.

The idea of view models is briefly touched on in https://templ.guide/core-concepts/view-models.

In your case I'd make a model that doesn't use the iterator directly, but instead, copies the results to a slice for display.

`templ x() {` components basically elide the implicit return. When Go gets `yield` statements, maybe it will be time to reconsider it, since there will be more options.

In a code component, you can do whatever you want, of course. 😀

**Comment by tarndt** [comment]
@a-h this made me look at the code that templ generates and I was disappointed to find its fully buffering the response rather than writing a the response writer (or a bufio.Writer around it). This means that I can't use templ for large streaming requests.

I also realized that
```
func ReleaseBuffer(b *bytes.Buffer) {
	b.Reset()
	bufferPool.Put(b)
}
```
doesn't check that b.Cap() is huge and intentionally not return it to the pool and allow it to be GC'd. 

Reading the code it looks like I can work around the last issue at least by passing my own *bytes.Buffer in as the destination. 

Thanks for the quick reply.

**Reply by a-h** [reply]
By the way, on the buffered response, see https://github.com/a-h/templ/pull/802 - it's not merged yet, but it switches generated code to using an underlying `bufio.Writer`.
